src/solana/solanaEngine.js

```diff
@@ -0,0 +1,390 @@
+      this.updateBalance(this.chainCode, balance.value.toString())
+    } catch (e) {
+      // fetching of account balances for uninitiated accounts throws error
+      if (
+        this.tokenCheckTransactionsStatus[this.chainCode] === 1 &&
+        this.transactionList[this.chainCode].length === 0
+      ) {
+        this.updateBalance(this.chainCode, '0')
+      }
+      this.error(`Error checking ${this.chainCode} address balance`, e)
+    }
+  }
+
+  async queryBlockheight() {
+    try {
+      const blockheight = asNumber(await this.fetchRpc('getSlot'))
+      if (blockheight > this.walletLocalData.blockHeight) {
+        this.walletLocalData.blockHeight = blockheight
+        this.walletLocalDataDirty = true
+        this.currencyEngineCallbacks.onBlockHeightChanged(
+          this.walletLocalData.blockHeight
+        )
+      }
+    } catch (e) {
+      this.error(`queryBlockheight Error `, e)
+    }
+  }
+
+  async queryFee() {
+    try {
+      const response = await this.fetchRpc('getRecentBlockhash')
+      const {
+        blockhash,
+        feeCalculator: { lamportsPerSignature }
+      } = asRecentBlockHash(response).value
+      this.feePerSignature = lamportsPerSignature.toString()
+      this.recentBlockhash = blockhash
+    } catch (e) {
+      this.error(`queryFee Error `, e)
+    }
+  }
+
+  processSolanaTransaction(tx: RpcGetTransaction, timestamp: number) {
+    const ourReceiveAddresses = []
+    const index = tx.transaction.message.accountKeys.findIndex(
+      account => account === this.keypair.publicKey.toBase58()
+    )
+    if (index < 0 || tx.meta == null) return
+    let amount = tx.meta.postBalances[index] - tx.meta.preBalances[index]
+    const fee = tx.meta.fee
+    if (amount < 0) {
+      amount -= fee
+    } else {
+      ourReceiveAddresses.push(this.keypair.publicKey.toBase58())
+    }
+    const edgeTransaction: EdgeTransaction = {
+      txid: tx.transaction.signatures[0],
+      date: timestamp,
+      currencyCode: this.chainCode,
+      blockHeight: tx.slot,
+      nativeAmount: amount.toString(),
+      networkFee: fee.toString(),
+      ourReceiveAddresses,
+      signedTx: ''
+    }
+    this.addTransaction(this.chainCode, edgeTransaction)
+  }
+
+  async queryTransactions() {
+    let before = null
+    const until =
+      this.otherData.newestTxid !== '' ? this.otherData.newestTxid : null
+    let txids = []
+    try {
+      // Gather all transaction IDs since we last updated
+      while (1) {
+        const params = [
+          this.keypair.publicKey.toBase58(),
+          {
+            until,
+            before,
+            limit: this.settings.txQueryLimit,
+            commitment: this.settings.commitment
+          }
+        ]
+        const response: RpcSignatureForAddress[] = await this.fetchRpc(
+          'getSignaturesForAddress',
+          params
+        )
+
+        // remove failed transactions from list
+        txids = txids.concat(
```

Comment:
Don't remove failed transactions if they incurred a fee to the user's wallet. Just like ETH, those are valid transactions, and the fee should still be shown. Set nativeamount to 0 and add it.